File: animator/src/main/java/com/richpathanimator/PathEvaluator.java
package com.richpathanimator;

import android.animation.TypeEvaluator;

import com.richpath.pathparser.PathDataNode;
import com.richpath.pathparser.PathParserCompat;

/**
 * Created by tarek on 7/14/17.
 */

public class PathEvaluator implements TypeEvaluator<PathDataNode[]> {

    private PathDataNode[] evaluatedNodes;

    @Override
    public PathDataNode[] evaluate(float fraction, PathDataNode[] startPathDataNodes, PathDataNode[] endPathDataNodes) {

        if (evaluatedNodes == null) {
            evaluatedNodes = PathParserCompat.deepCopyNodes(startPathDataNodes);
        }

        for (int i = 0; i < startPathDataNodes.length; i++) {
            for (int j = 0; j < startPathDataNodes[i].getParams().length; j++) {
                float startFloat = startPathDataNodes[i].getParams()[j];
                float value = startFloat + fraction * (endPathDataNodes[i].getParams()[j] - startFloat);
                evaluatedNodes[i].getParams()[j] = value;
            }
        }

        return evaluatedNodes;
    }


}

File: animator/src/main/kotlin/com/richpathanimator/PathEvaluator.kt
package com.richpathanimator

import android.animation.TypeEvaluator
import com.richpath.pathparser.PathDataNode
import com.richpath.pathparser.PathParserCompat

class PathEvaluator: TypeEvaluator<Array<PathDataNode>?> {
    private var evaluatedNodes: Array<PathDataNode>? = null

    override fun evaluate(fraction: Float, startPathDataNodes: Array<PathDataNode>?, endPathDataNodes: Array<PathDataNode>?): Array<PathDataNode>? {
        if (startPathDataNodes == null || endPathDataNodes == null) return null
        val evaluatedNodes = this.evaluatedNodes ?: PathParserCompat.deepCopyNodes(startPathDataNodes)
        this.evaluatedNodes = evaluatedNodes

        val startNodeSize = startPathDataNodes.size
        for (i in 0 until startNodeSize) {
            val nodeParamSize = startPathDataNodes[i].params.size
            for (j in 0 until nodeParamSize) {
                val startFloat = startPathDataNodes[i].params[j]
                val value = startFloat + fraction * (endPathDataNodes[i].params[j] - startFloat)
                evaluatedNodes[i].params[j] = value
            }
        }
        return evaluatedNodes
    }
}
